File: Level 2/Module9 - Stacks, Queues, Deques/MinStack.cpp
//**********************************************************************************
// pbm : https://leetcode.com/problems/min-stack/description/
//**********************************************************************************
#include <bits/stdc++.h>
#include <stack>
using namespace std;

class MinStack {
public:
    // Main stack stores all values.
    stack<int> s;
    // pre stores the running minimums in the same order as s.
    stack<int> pre;

    MinStack() {
        // Default constructor
    }

    void push(int val) {
        s.push(val);

        // If the new value is smaller or equal to current minimum,
        // store it in the min-stack.
        if (pre.empty() || val <= pre.top()) {
            pre.push(val);
        }
    }

    void pop() {
        // If the popped value is the current minimum,
        // also remove it from the min-stack.
        if (!s.empty() && !pre.empty() && s.top() == pre.top()) {
            pre.pop();
        }
        s.pop();
    }

    int top() {
        return s.top();
    }

    int getMin() {
        return pre.top();
    }
};

File: Level 2/Module9 - Stacks, Queues, Deques/MinStack.java
import java.util.Stack;
/*

    Min stack operation needs to run in O(1) time. Binary search tree and heap data structures
    are often great for keeping track of minimum, their core operations find, add and remove
    are log(n) time complexity. So we need to use some other data structure to keep track of minimum in O(1) time.

*/
/*
    Approach 1: Stack of Value and Minimum Pair

*/

class MinstackApproach1 {

    private Stack<int[]> stack;
    
    MinstackApproach1() {
        stack = new Stack<>();
    }

    public void push(int val) {
        if(stack.isEmpty()) {
            stack.push(new int[]{val, val});
        } else {
            stack.push(new int[]{val, Math.min(val, stack.peek()[1])});
        }
    }

    public void pop() {
        if(!stack.isEmpty()) {
            stack.pop();
        }
    }

    public int top() {
        if(!stack.isEmpty()) {
            return stack.peek()[0];
        }
        throw new RuntimeException("Stack is empty");
    }

    public int getMin() {
        if(!stack.isEmpty()) {
            return stack.peek()[1];
        }
        throw new RuntimeException("Stack is empty");
    }
}

/*
    Approach 2: Two Stacks with frequency count of minimum element


*/

public class MinStack {
    private Stack<Integer> stack;
    private Stack<int[]> minStack;

    public MinStack() {
        stack = new Stack<>();
        minStack = new Stack<>();
    }

    public void push(int val) {
        stack.push(val);
        if (minStack.isEmpty() || val < minStack.peek()[0]) {
            minStack.push(new int[]{val, 1});
        } else if (val == minStack.peek()[0]) {
            minStack.peek()[1]++;
        }
    }

    public void pop() {
        if (!stack.isEmpty()) {
            int val = stack.pop();
            if (val == minStack.peek()[0]) {
                if (minStack.peek()[1] > 1) {
                    minStack.peek()[1]--;
                } else {
                    minStack.pop();
                }
            }
        }
    }

    public int top() {
        if (!stack.isEmpty()) {
            return stack.peek();
        }
        throw new RuntimeException("Stack is empty");
    }

    public int getMin() {
        if (!minStack.isEmpty()) {
            return minStack.peek()[0];
        }
        throw new RuntimeException("Stack is empty");
    }
}
